zkevm-circuits/src/evm_circuit/op_execution.rs

```diff
@@ -0,0 +1,524 @@
+use super::{
+    param::{
+        CIRCUIT_HEIGHT, CIRCUIT_WIDTH, NUM_CELL_OP_EXECTION_STATE,
+        NUM_CELL_OP_GADGET_SELECTOR, NUM_CELL_RESUMPTION,
+    },
+    Case, Cell, Constraint, CoreStateInstance, ExecutionStep, Lookup, Word,
+};
+use crate::util::Expr;
+use bus_mapping::evm::OpcodeId;
+use halo2::{
+    arithmetic::FieldExt,
+    circuit::Region,
+    plonk::{ConstraintSystem, Error, Expression},
+};
+use std::{collections::HashMap, ops::Range};
+
+mod arithmetic;
+mod push;
+
+use arithmetic::AddGadget;
+use push::PushGadget;
+
+fn bool_switches_constraints<F: FieldExt>(
+    bool_switches: &[Cell<F>],
+) -> Vec<Expression<F>> {
+    let mut constraints = Vec::with_capacity(bool_switches.len() + 1);
+    let mut sum_to_one = 0.expr();
+
+    for switch in bool_switches {
+        constraints.push(switch.expr() * (1.expr() - switch.expr()));
+        sum_to_one = sum_to_one + switch.expr();
+    }
+
+    constraints.push(1.expr() - sum_to_one);
+
+    constraints
+}
+
+#[derive(Debug)]
+struct CaseConfig {
+    case: Case,
+    num_word: usize,
+    num_cell: usize,
+    will_halt: bool,
+}
+
+impl CaseConfig {
+    fn num_total_cell(&self) -> usize {
+        32 * self.num_word
+            + self.num_cell
+            + if self.will_halt {
+                NUM_CELL_RESUMPTION
+            } else {
+                0
+            }
+    }
+
+    // allocate indexes of cells for words, cells and unused. It assumes input
+    // free_cells are in order of rotation and then column index.
+    fn allocation_idxs<F: FieldExt>(
```

Comment:
Just a better function name suggestion. 

```suggestion
    fn allocate_indices<F: FieldExt>(
```

Reply:
Thanks, `allocate_indices` sounds better. Updated in [`ce345d5`](https://github.com/appliedzkp/zkevm-circuits/pull/26/commits/ce345d5b26172555e678910b2e36c09389f022f9).